moshi/src/main/java/com/squareup/moshi/ClassJsonAdapter.kt

```diff
@@ -0,0 +1,229 @@
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import com.squareup.moshi.internal.isKotlin
+import com.squareup.moshi.internal.isPlatformType
+import com.squareup.moshi.internal.jsonAnnotations
+import com.squareup.moshi.internal.jsonName
+import com.squareup.moshi.internal.resolve
+import com.squareup.moshi.internal.rethrowCause
+import java.lang.reflect.Field
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.Modifier
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+
+/**
+ * Emits a regular class as a JSON object by mapping Java fields to JSON object properties.
+ *
+ * <h1>Platform Types</h1>
+ *
+ * Fields from platform classes are omitted from both serialization and deserialization unless they
+ * are either public or protected. This includes the following packages and their subpackages:
+ *
+ *  * android.*
+ *  * androidx.*
+ *  * java.*
+ *  * javax.*
+ *  * kotlin.*
+ *  * kotlinx.*
+ *  * scala.*
+ *
+ */
+internal class ClassJsonAdapter<T>(
+  private val classFactory: ClassFactory<T>,
+  fieldsMap: Map<String, FieldBinding<*>>
+) : JsonAdapter<T>() {
+  private val fieldsArray = fieldsMap.values.toTypedArray()
+  private val options = JsonReader.Options.of(*fieldsMap.keys.toTypedArray())
+
+  override fun fromJson(reader: JsonReader): T {
+    val result: T = try {
+      classFactory.newInstance()
+    } catch (e: InstantiationException) {
+      throw RuntimeException(e)
+    } catch (e: InvocationTargetException) {
+      throw e.rethrowCause()
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+
+    try {
+      reader.beginObject()
+      while (reader.hasNext()) {
+        val index = reader.selectName(options)
+        if (index == -1) {
+          reader.skipName()
+          reader.skipValue()
+          continue
+        }
+        fieldsArray[index].read(reader, result as Any)
+      }
+      reader.endObject()
+      return result
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+  }
+
+  override fun toJson(writer: JsonWriter, value: T?) {
+    try {
+      writer.beginObject()
+      fieldsArray.forEach { fieldBinding ->
+        writer.name(fieldBinding.name)
+        fieldBinding.write(writer, value as Any)
+      }
+      writer.endObject()
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+  }
+
+  override fun toString() = "JsonAdapter($classFactory)"
+
+  internal open class FieldBinding<T>(val name: String, val field: Field, val adapter: JsonAdapter<T>) {
```

Comment:
Why is this `open` Same with the functions in it?

Reply:
I marked to `open` to satisfy japicmp. Should I add an exclusion since the class is internal?

Reply:
yeah exclude it, it's package-private and not intended to be subclassed

Reply:
https://github.com/square/moshi/pull/1470/commits/db568ec70896d16a6f0a99063ce4f8a962c3b66b